Content.Shared/Decals/SharedDecalSystem.cs

```diff
@@ -0,0 +1,153 @@
+        {
+            var comp = EntityManager.EnsureComponent<DecalGridComponent>(MapManager.GetGrid(msg.GridId).GridEntityId);
+            ChunkIndex[msg.GridId] = new();
+            foreach (var (indices, decals) in comp.ChunkCollection.ChunkCollection)
+            {
+                foreach (var uid in decals.Keys)
+                {
+                    ChunkIndex[msg.GridId][uid] = indices;
+                }
+            }
+        }
+
+        protected DecalGridComponent.DecalGridChunkCollection DecalGridChunkCollection(GridId gridId) => EntityManager
+            .GetComponent<DecalGridComponent>(MapManager.GetGrid(gridId).GridEntityId).ChunkCollection;
+        protected Dictionary<Vector2i, Dictionary<uint, Decal>> ChunkCollection(GridId gridId) => DecalGridChunkCollection(gridId).ChunkCollection;
+
+        protected virtual void DirtyChunk(GridId id, Vector2i chunkIndices) {}
+
+        protected bool RemoveDecalInternal(GridId gridId, uint uid)
+        {
+            if (!RemoveDecalHook(gridId, uid)) return false;
+
+            if (!ChunkIndex.TryGetValue(gridId, out var values) || !values.TryGetValue(uid, out var indices))
+            {
+                return false;
+            }
+
+            var chunkCollection = ChunkCollection(gridId);
+            if (!chunkCollection.TryGetValue(indices, out var chunk) || !chunk.Remove(uid))
+            {
+                return false;
+            }
+
+            if (chunkCollection[indices].Count == 0)
+                chunkCollection.Remove(indices);
+
+            ChunkIndex[gridId]?.Remove(uid);
+            DirtyChunk(gridId, indices);
+            return true;
+        }
+
+        protected virtual bool RemoveDecalHook(GridId gridId, uint uid) => true;
+
+        private (Box2 view, MapId mapId) CalcViewBounds(in EntityUid euid)
+        {
+            var xform = EntityManager.GetComponent<TransformComponent>(euid);
+
+            var view = Box2.UnitCentered.Scale(_viewSize).Translated(xform.WorldPosition);
+            var map = xform.MapID;
+
+            return (view, map);
+        }
+
+        protected Dictionary<GridId, HashSet<Vector2i>> GetChunksForViewers(HashSet<EntityUid> viewers)
+        {
+            var chunks = new Dictionary<GridId, HashSet<Vector2i>>();
+            foreach (var viewerUid in viewers)
+            {
+                var (bounds, mapId) = CalcViewBounds(viewerUid);
+                foreach (var grid in MapManager.FindGridsIntersecting(mapId, bounds, approx: true))
+                {
+                    if(!chunks.ContainsKey(grid.Index))
+                        chunks[grid.Index] = new();
+                    var enumerator = new ChunkIndicesEnumerator(grid.InvWorldMatrix.TransformBox(bounds), ChunkSize);
+                    while (enumerator.MoveNext(out var indices))
+                    {
+                        chunks[grid.Index].Add(indices.Value);
+                    }
+                }
+            }
```

Comment:
There's an enumerator for FindGridsIntersecting if you want to get fancy

Reply:
what are the upsides of using an enumerator like that?

Reply:
Allocations